lib/framework/NetworkSettingsService.cpp

```diff
@@ -83,31 +86,364 @@ void NetworkSettingsService::manageSTA() {
+        }
 #endif
-        // }
     } else { // not connected but STA-mode active => disconnect
         reconfigureWiFiConnection();
     }
 }
 
-// handles if wifi stopped
-void NetworkSettingsService::WiFiEvent(WiFiEvent_t event) {
-    if (event == ARDUINO_EVENT_WIFI_STA_STOP) {
+// set the TxPower based on the RSSI (signal strength), picking the lowest value
+// code is based of RSSI (signal strength) and copied from Tasmota's WiFiSetTXpowerBasedOnRssi() which is copied ESPEasy's ESPEasyWifi.SetWiFiTXpower() function
+void NetworkSettingsService::setWiFiPowerOnRSSI() {
+    // Range ESP32  : 2dBm - 20dBm
+    // 802.11b - wifi1
+    // 802.11a - wifi2
+    // 802.11g - wifi3
+    // 802.11n - wifi4
+    // 802.11ac - wifi5
+    // 802.11ax - wifi6
+
+    int max_tx_pwr = MAX_TX_PWR_DBM_n;        // assume wifi4
+    int threshold  = WIFI_SENSITIVITY_n + 30; // Margin in dBm * 10 on top of threshold
+
+    // Assume AP sends with max set by ETSI standard.
+    // 2.4 GHz: 100 mWatt (20 dBm)
+    // US and some other countries allow 1000 mW (30 dBm)
+    int rssi    = WiFi.RSSI() * 10;
+    int newrssi = rssi - 200; // We cannot send with over 20 dBm, thus it makes no sense to force higher TX power all the time.
+
+    int min_tx_pwr = 0;
+    if (newrssi < threshold) {
+        min_tx_pwr = threshold - newrssi;
+    }
+    if (min_tx_pwr > max_tx_pwr) {
+        min_tx_pwr = max_tx_pwr;
+    }
+
+    uint8_t set_power = min_tx_pwr / 10; // this is the recommended power setting to use
+
+    // from WiFIGeneric.h use:
+    //  WIFI_POWER_19_5dBm = 78,// 19.5dBm
+    //  WIFI_POWER_19dBm = 76,// 19dBm
+    //  WIFI_POWER_18_5dBm = 74,// 18.5dBm
+    //  WIFI_POWER_17dBm = 68,// 17dBm
+    //  WIFI_POWER_15dBm = 60,// 15dBm
+    //  WIFI_POWER_13dBm = 52,// 13dBm
+    //  WIFI_POWER_11dBm = 44,// 11dBm
+    //  WIFI_POWER_8_5dBm = 34,// 8.5dBm
+    //  WIFI_POWER_7dBm = 28,// 7dBm
+    //  WIFI_POWER_5dBm = 20,// 5dBm
+    //  WIFI_POWER_2dBm = 8,// 2dBm
+    //  WIFI_POWER_MINUS_1dBm = -4// -1dBm
+    wifi_power_t p = WIFI_POWER_2dBm;
+    if (set_power > 19)
+        p = WIFI_POWER_19_5dBm;
+    else if (set_power > 18)
+        p = WIFI_POWER_18_5dBm;
+    else if (set_power >= 17)
+        p = WIFI_POWER_17dBm;
+    else if (set_power >= 15)
+        p = WIFI_POWER_15dBm;
+    else if (set_power >= 13)
+        p = WIFI_POWER_13dBm;
+    else if (set_power >= 11)
+        p = WIFI_POWER_11dBm;
+    else if (set_power >= 8)
+        p = WIFI_POWER_8_5dBm;
+    else if (set_power >= 7)
+        p = WIFI_POWER_7dBm;
+    else if (set_power >= 5)
+        p = WIFI_POWER_5dBm;
+
+#ifdef EMSESP_DEBUG
+    emsesp::EMSESP::logger().debug("Recommended set WiFi Tx Power (set_power %d, new power %d, rssi %d, threshold %d", set_power, p, rssi, threshold);
+#else
+    char result[10];
+    emsesp::EMSESP::logger().info("Setting WiFi Tx Power to %s dBm", emsesp::Helpers::render_value(result, (double)(p / 4), 1));
```

Comment:
Wrong bracket, => ((double)p) / 4 to get the x.5 settings

Reply:
fixed.